clang/lib/CConv/ProgramInfo.cpp

```diff
@@ -513,56 +475,94 @@ void ProgramInfo::specialCaseVarIntros(ValueDecl *D, ASTContext *Context) {
-        PVC->constrainToWild(CS, Rsn, &PL);
-      }
-    }
+    ConstraintVariable *CV = getVariable(D, Context).getValueOr(nullptr);
+    if (PVConstraint *PVC = dyn_cast_or_null<PVConstraint>(CV))
+      PVC->constrainToWild(CS, Rsn, &PL);
   }
 }
 
 // For each pointer type in the declaration of D, add a variable to the
 // constraint system for that pointer type.
 void ProgramInfo::addVariable(clang::DeclaratorDecl *D,
-                              clang::ASTContext *astContext) {
+                              clang::ASTContext *AstContext) {
   assert(!persisted);
 
-  PersistentSourceLoc PLoc = PersistentSourceLoc::mkPSL(D, *astContext);
+  PersistentSourceLoc PLoc = PersistentSourceLoc::mkPSL(D, *AstContext);
   assert(PLoc.valid());
 
-  // We only add a PVConstraint or an FVConstraint if the set at
-  // Variables[PLoc] does not contain one already. TODO: Explain why would this happen
-  CVarSet &S = Variables[PLoc];
-  if (S.size()) return;
+  // We only add a PVConstraint if Variables[PLoc] does not exist.
+  // Functions are exempt from this check because they need to be added to the
+  // Extern/Static function map even if they are inside a macro expansion.
+  if (Variables.find(PLoc) != Variables.end() && !isa<FunctionDecl>(D)) {
+    // Two variables can have the same source locations when they are
+    // declared inside the same macro expansion. The first instance of the
+    // source location will have been constrained to WILD, so it's safe to bail
+    // without doing anymore work.
+    if (!Rewriter::isRewritable(D->getLocation())) {
+      // If we're not in a macro, we should make the constraint variable WILD
+      // anyways. This happens if the name of the variable is a macro defined
+      // differently is different parts of the program.
+      std::string Rsn = "Duplicate source location. Possibly part of a macro.";
+      Variables[PLoc]->constrainToWild(CS, Rsn);
+    }
+    return;
+  }
+
+  ConstraintVariable *NewCV = nullptr;
 
   if (FunctionDecl *FD = dyn_cast<FunctionDecl>(D)) {
     // Function Decls have FVConstraints.
-    FVConstraint *F = new FVConstraint(D, *this, *astContext);
+    FVConstraint *F = new FVConstraint(D, *this, *AstContext);
     F->setValidDecl();
+
+    // Handling of PSL collision for functions is different since we need to
+    // consider the static and extern function maps.
+    if (Variables.find(PLoc) != Variables.end()) {
+      // Try to find a previous definition based on function name
+      if (!getFuncConstraint(FD, AstContext)) {
+        constrainWildIfMacro(F, FD->getLocation());
+        insertNewFVConstraint(FD, F, AstContext);
+      } else {
+        // FIXME: Visiting same function in same source location twice.
+        //        This shouldn't happen, but it does for some std lib functions
+        //        on our benchmarks programs (vsftpd, lua, etc.). Bailing for
+        //        now, but a real fix will catch and prevent this earlier.
+      }
```

Comment:
Hm, so is this something still to figure out?

Reply:
It's still a bit odd. The current solution works and avoids the duplicate function definition warnings, but I'm not fully happy with it. If we can make a change so that the duplicate definitions never reach this point, I'd be happier with it, but as is, it's correct and gives no extraneous warnings.

It might also be the case that there really are duplicate function definitions in the standard library, perhaps via define trickery in header files. It'd be good to know if this is what's happening, but if it is, the current code is still correct.

Reply:
OK that's a good point.